File: server/lib/comcent/transcript.ex
defmodule Comcent.Transcript do
  @moduledoc """
  Module for handling transcript creation and formatting.
  """

  alias Comcent.Schemas.CallStory

  @type sentence :: %{
          current_party: String.t(),
          timestamp: integer(),
          text: String.t()
        }

  @type chat :: %{
          current_party: String.t(),
          start: number(),
          message: String.t()
        }

  @doc """
  Gets sorted sentences from call transcripts.
  """
  @spec get_sorted_sentences(CallStory.t()) :: [sentence()]
  def get_sorted_sentences(call_story) do
    sentences =
      Enum.flat_map(call_story.call_transcripts, fn transcript ->
        data = transcript.transcript_data

        related_call_span =
          Enum.find(call_story.call_spans, fn span ->
            metadata = span.metadata

            span.current_party == transcript.current_party &&
              span.type == "RECORDING" &&
              metadata["direction"] == "in"
          end)

        start_at = related_call_span && related_call_span.start_at

        channels = get_in(data, ["results", "channels"]) || []

        Enum.flat_map(channels, fn channel ->
          alternatives = channel["alternatives"] || []

          Enum.flat_map(alternatives, fn alternative ->
            words = alternative["words"] || []

            Enum.map(words, fn word ->
              timestamp =
                if start_at do
                  # Convert start_at to Unix timestamp if it's a DateTime
                  start_at_unix =
                    case start_at do
                      %DateTime{} -> DateTime.to_unix(start_at)
                      _ -> start_at
                    end

                  # Add the word's start time to the call start time
                  start_at_unix + trunc(word["start"])
                else
                  0
                end

              %{
                current_party: transcript.current_party,
                timestamp: timestamp,
                text: word["word"]
              }
            end)
          end)
        end)
      end)

    Enum.sort_by(sentences, & &1.timestamp)
  end

  @doc """
  Creates a formatted text transcript from call story.
  """
  @spec create_transcript_text(CallStory.t()) :: {:ok, String.t()}
  def create_transcript_text(call_story) do
    sorted_sentences = get_sorted_sentences(call_story)

    {final_transcript, _} =
      Enum.reduce(sorted_sentences, {"", ""}, fn sentence, {transcript, current_party} ->
        if current_party != sentence.current_party do
          new_transcript = transcript <> "\n\n#{sentence.current_party}:"
          {new_transcript, sentence.current_party}
        else
          {transcript <> " " <> sentence.text, current_party}
        end
      end)

    {:ok, String.trim(final_transcript)}
  end

  @doc """
  Creates a chat-style transcript from call story.
  """
  @spec create_transcript_chat(CallStory.t()) :: [chat()]
  def create_transcript_chat(call_story) do
    sorted_sentences = get_sorted_sentences(call_story)

    {transcript_messages, current_chat} =
      Enum.reduce(sorted_sentences, {[], nil}, fn sentence, {messages, current_chat} ->
        if current_chat == nil do
          new_chat = %{
            current_party: sentence.current_party,
            start: sentence.timestamp,
            message: sentence.text
          }

          {messages, new_chat}
        else
          if current_chat.current_party != sentence.current_party do
            new_chat = %{
              current_party: sentence.current_party,
              start: sentence.timestamp,
              message: sentence.text
            }

            {messages ++ [current_chat], new_chat}
          else
            updated_chat = %{
              current_chat
              | message: current_chat.message <> " " <> sentence.text
            }

            {messages, updated_chat}
          end
        end
      end)

    if current_chat do
      transcript_messages ++ [current_chat]
    else
      transcript_messages
    end
  end
end

File: packages/web-app/src/lib/server/transcripts/transcript.ts
import type { CallSpan, CallStory, CallTranscript } from '$lib/types/database';
import type { DeepgramResult } from '../types/DeepgramResult.js';
import { logger } from '../logger.js';
import { RecordingMetadata } from '../types/RecordingMetadata.js';
import moment from 'moment';

type Sentence = {
  currentParty: string;
  start: number;
  text: string;
};

type CallStoryWithTranscripts = CallStory & {
  callTranscripts: CallTranscript[];
  callSpans: CallSpan[];
};

function getSortedSentences(callStory: CallStoryWithTranscripts): Sentence[] {
  const sentences = [];
  for (const t of callStory.callTranscripts) {
    const data = t.transcriptData as DeepgramResult;
    const relatedCallSpan = callStory.callSpans.find((span) => {
      const metadata = span.metadata as any as RecordingMetadata;
      return (
        span.currentParty === t.currentParty &&
        span.type === 'RECORDING' &&
        metadata.direction === 'in'
      );
    });
    const startAt = relatedCallSpan?.startAt;
    const paragraphs = data.results.channels[0].alternatives[0]?.paragraphs?.paragraphs;
    if (!paragraphs) {
      logger.info(
        `No paragraphs found in transcript for call story ${t.callStoryId} recordingSpanId ${t.recordingSpanId}`,
      );
      continue;
    }
    for (const paragraph of paragraphs) {
      for (const sentence of paragraph.sentences) {
        sentences.push({
          currentParty: t.currentParty,
          timestamp: moment(startAt).add(sentence.start, 'seconds').unix(),
          ...sentence,
        });
      }
    }
  }
  return sentences.sort((a, b) => a.timestamp - b.timestamp);
}

type Chat = {
  currentParty: string;
  start: number;
  message: string;
};

export function createTranscriptChat(callStory: CallStoryWithTranscripts) {
  const sortedSentences = getSortedSentences(callStory);
  const transcriptMessages = [];
  let currentChat: Chat | null = {
    currentParty: sortedSentences?.[0]?.currentParty,
    start: sortedSentences?.[0]?.start,
    message: '',
  };
  for (const sentence of sortedSentences) {
    if (currentChat?.currentParty !== sentence.currentParty) {
      currentChat.message = currentChat.message.trim();
      transcriptMessages.push(currentChat);
      currentChat = {
        currentParty: sentence.currentParty,
        start: sentence.start,
        message: '',
      };
    }
    currentChat.message += ` ${sentence.text}`;
  }
  currentChat.message = currentChat.message.trim();
  if (currentChat?.message) {
    transcriptMessages.push(currentChat);
  }

  return transcriptMessages;
}
